lib/kernel/src/application_controller.erl

```diff
@@ -343,9 +343,9 @@ get_pid_env(Master, Key) ->
     end.
 
 get_env(AppName, Key) ->
-    case ets:lookup(ac_tab, {env, AppName, Key}) of
-	[{_, Val}] -> {ok, Val};
-	_ -> undefined
+    case ets:lookup_element(ac_tab, {env, AppName, Key}, 2, not_found) of
+        not_found -> undefined,
+        Val -> {ok, Val}
```

Comment:
This (apart from the already mentioned issue with tuples of size >2) is pretty questionable. If `not_found` is stored in the table, it is indistinguishable from there being no record at all and `undefined` will be returned. I suggest to create a reference beforehand, use it as the default, and match on that.

Reply:
I agree that a reference should be used, but since creating a reference does not come for free, I think you should do some benchmarking whether it would be an improvement compared to the original code.

This table is writable by all processes in the system, so there is a (very) slight possibility that testing the tuple size could detect a garbage value written by another process. However, since this is a system process always running, I think that (very) slightly improving performance is worth more than keeping the tuple test that is very unlikely to go wrong.

Reply:
I've just done some benchmarking.
Running `erlperf -l 10K --init 'application:set_env(foobar, key, 42).' 'application:get_env(foobar, key).'` 10 times for each configuration I get the following results:
Baseline: 15314 Ki to 15649 Ki QPS
With this patch: 18939 Ki to 21598 Ki QPS (most values around 21000)
Using a reference instead of an atom: 17391 Ki to 19531 Ki (most values around 18000)

I then checked whether using `{key, 0}` as the key instead of just the atom `key` changes anything, and I got the following results:
Baseline: 12225 Ki to 12903 Ki
With this patch: 18018 Ki to 18727 Ki
Using a reference instead of an atom: 15106 Ki to 17361 Ki (most values around 16000)

So lookup_element + make_ref seems to always be faster than lookup, but the make_ref is far from free. Would sticking to using an atom and just making it more unique (something like `not_found_application_controller_get_env`) be considered safe enough?